### PR TITLE
Bug 1985997: readiness checks handle network errors better

### PR DESCRIPTION
**Pull request by p0lyn0mial**

### CODE REVIEW

File: pkg/operator/startupmonitorreadiness/readiness_checks.go [hunk 1]
```diff
@@ -275,11 +276,20 @@ func goodHealthzEtcdEndpoint(client *http.Client, rawURL string) func(context.Co
 func doHTTPCheckAndTransform(ctx context.Context, client *http.Client, rawURL string, checkName string, httpCheckFn func(ctx context.Context, client *http.Client, rawURL string) (int, string, error)) (bool, string, string) {
 	statusCode, response, err := httpCheckFn(ctx, client, rawURL)
 	if err != nil {
+		if utilnet.IsConnectionRefused(err) {
+			return false, "ConnectionRefused", fmt.Sprintf("kube-apiserver is not reachable: %v", err)
```
Comment (sttts) [inline pkg/operator/startupmonitorreadiness/readiness_checks.go:280]: would unify all of these into "APIError".

Then this one: "waiting for kube-apiserver to listen on port :6443"

File: pkg/operator/startupmonitorreadiness/readiness_checks.go [hunk 2]
```diff
@@ -275,11 +276,20 @@ func goodHealthzEtcdEndpoint(client *http.Client, rawURL string) func(context.Co
...
+		}
+		if utilnet.IsConnectionReset(err) || utilnet.IsProbableEOF(err) {
+			return false, "ConnectionError", fmt.Sprintf("failed sending request to kube-apiserver: %v", err)
```
Comment (sttts) [inline pkg/operator/startupmonitorreadiness/readiness_checks.go:283]: "waiting for kube-apiserver static pod %s to answer requests: %v"

File: pkg/operator/startupmonitorreadiness/readiness_checks.go [hunk 3]
```diff
@@ -275,11 +276,20 @@ func goodHealthzEtcdEndpoint(client *http.Client, rawURL string) func(context.Co
...
+		}
+		if utilnet.IsTimeout(err) {
+			return false, "Timeout", fmt.Sprintf("no response in the allotted time from kube-apiserver: %v", err)
```
Comment (sttts) [inline pkg/operator/startupmonitorreadiness/readiness_checks.go:286]: "request to kube-apiserver static pod timed out: %v"

File: pkg/operator/startupmonitorreadiness/readiness_checks.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/operator/startupmonitorreadiness/readiness_checks.go:280]: won't `APIError` imply we got an error from the server?

File: pkg/operator/startupmonitorreadiness/readiness_checks.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/operator/startupmonitorreadiness/readiness_checks.go:283]: mhm, we are not waiting, we failed possibly due to some network errror

Comment (sttts) [reply pkg/operator/startupmonitorreadiness/readiness_checks.go:283]: the whole startup-monitor is waiting. I would formulate these in that way. Maybe this one is different as it is a random error. But for connection refused we are waiting for the process and port.

File: pkg/operator/startupmonitorreadiness/readiness_checks.go (hunk 1 above)
Comment (aojea) [reply pkg/operator/startupmonitorreadiness/readiness_checks.go:280]: network errors are a thing in golang
```
 if netError, ok := err.(net.Error)
```
they also have a Timeout() and Temporary() method to help identify them

File: pkg/operator/startupmonitorreadiness/readiness_checks.go [hunk 4]
```diff
@@ -279,11 +276,20 @@ func goodHealthzEtcdEndpoint(client *http.Client, rawURL string) func(context.Co
 func doHTTPCheckAndTransform(ctx context.Context, client *http.Client, rawURL string, checkName string, httpCheckFn func(ctx context.Context, client *http.Client, rawURL string) (int, string, error)) (bool, string, string) {
 	statusCode, response, err := httpCheckFn(ctx, client, rawURL)
 	if err != nil {
+		if utilnet.IsConnectionRefused(err) {
+			return false, "NetworkError", fmt.Sprintf("waiting for kube-apiserver static pod to listen on port 6443: %v", err)
+		}
+		if utilnet.IsConnectionReset(err) || utilnet.IsProbableEOF(err) {
+			return false, "NetworkError", fmt.Sprintf("failed sending request to kube-apiserver: %v", err)
```
Comment (aojea) [inline pkg/operator/startupmonitorreadiness/readiness_checks.go:283]: is connection reset  is that someone RSTed the connection, probably misbehaving loadbalancer

File: pkg/operator/startupmonitorreadiness/readiness_checks.go [hunk 5]
```diff
@@ -279,11 +276,20 @@ func goodHealthzEtcdEndpoint(client *http.Client, rawURL string) func(context.Co
 func doHTTPCheckAndTransform(ctx context.Context, client *http.Client, rawURL string, checkName string, httpCheckFn func(ctx context.Context, client *http.Client, rawURL string) (int, string, error)) (bool, string, string) {
 	statusCode, response, err := httpCheckFn(ctx, client, rawURL)
 	if err != nil {
+		if utilnet.IsConnectionRefused(err) {
+			return false, "NetworkError", fmt.Sprintf("waiting for kube-apiserver static pod to listen on port 6443: %v", err)
```
Comment (aojea) [inline pkg/operator/startupmonitorreadiness/readiness_checks.go:280]: is this only for SNO?

File: pkg/operator/startupmonitorreadiness/readiness_checks.go (hunk 4 above)
Comment (p0lyn0mial) [reply pkg/operator/startupmonitorreadiness/readiness_checks.go:283]: yeah - probably

File: pkg/operator/startupmonitorreadiness/readiness_checks.go (hunk 5 above)
Comment (p0lyn0mial) [reply pkg/operator/startupmonitorreadiness/readiness_checks.go:280]: yes - `readiness_checks` apply on an `sno` cluster

Comment (sttts) [reply pkg/operator/startupmonitorreadiness/readiness_checks.go:280]: technically this can run in HA too.

**Pull request merged**